crates/rust-analyzer/src/global_state.rs

```diff
@@ -365,6 +412,54 @@ impl GlobalState {
 
         self.analysis_host.apply_change(change);
 
+        let config_change = {
+            let mut change = ConfigChange::default();
+            let snap = self.analysis_host.analysis();
+
+            for (file_id, (vfs_path, text)) in ratoml_text_map {
+                // If change has been made to a ratoml file that
+                // belongs to a non-local source root, we will ignore it.
+                // As it doesn't make sense a users to use external config files.
+                if let Ok(source_root) = snap.source_root_id(file_id) {
+                    if let Ok(true) = snap.is_local_source_root(source_root) {
+                        if let Some((old_file, old_path, old_text)) =
+                            change.change_ratoml(source_root, file_id, vfs_path.clone(), text)
+                        {
+                            // SourceRoot has more than 1 RATOML files. In this case lexicographically smaller wins.
+                            if old_path < vfs_path {
+                                warn!(?old_path, ?vfs_path, "Two `rust-analyzer.toml` files were found inside the same crate. {vfs_path} has no effect.");
+                                // Put the old one back in.
+                                change.change_ratoml(source_root, old_file, old_path, old_text);
+                            }
+                        }
+                    }
+                } else {
+                    // Mapping to a SourceRoot should always end up in `Ok`
+                    error!("Mapping to SourceRootId failed.");
+                }
+            }
+
+            if let Some((file_id, Some(txt))) = user_config_file {
+                change.change_user_config(Some((file_id, txt)));
+            }
+
+            if let Some((file_id, Some(txt))) = root_path_ratoml {
+                change.change_root_ratoml(Some((file_id, txt)));
+            }
+
+            change
+        };
+
+        let mut error_sink = ConfigError::default();
+        let config = self.config.apply_change(config_change, &mut error_sink);
+
+        if config.should_update() {
```

Comment:
We should return the `should_update` bool when calling `apply_change` instead of encoding it in the config as a field